microscope/devices.py

```diff
@@ -481,8 +483,12 @@ def _fetch_data(self):
         return None
 
     def _process_data(self, data):
-        """Do any data processing and return data."""
-        return data
+        """Do any data processing and return data.
+
+        Subclasses should call super()._process_data(data) after doing their
+        own processing."""
+        import functools
+        return functools.reduce(lambda x, f: f(x), self.pipeline, data)
```

Comment:
If we have an attribute which is a list of processing steps, why should subclasses do their own processing first instead of just appending to `self.pipeline`?

Reply:
Yes - I think we should do one or the other. This requires some thought, though.

Originally ```_process_data``` was a method on ```DataDevice```s to implement the processing step, and I think that was meant to be pure virtual.

Now we have ```Devices``` that are not ```DataDevices``` that need to add processing steps, and I addressed that with the pipeline, but that needs ```super()._process_data()``` to ensure that the pipeline is executed after the ```DataDevice``` does anything it needs to do (e.g. running it through a LUT, type conversions, reshaping arrays). 

The ```DataDevice``` could just add it's processing to the start of the pipeline, instead, and ```_process_data``` would be implemented on ```DataDevice``` and just process the pipeline.

Reply:
We talked about this at length. 
* Most classes derived from ```DataDevice``` should probably add to the pipeline rather than override ```_process_data```.
* Devices that override ```_process_data``` must call ```super()._process_data()```.
* The docstring on ```DataDevice._process_data``` should indicate this clearly.
* We could make ```pipeline``` private, but that would require methods to add/remove steps from it.

Reply:
See comments in review immediately below this one.